libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/http/FastBrowserJsHttpEngine.kt

```diff
@@ -0,0 +1,95 @@
+package com.apollographql.apollo3.network.http
+
+import com.apollographql.apollo3.api.http.HttpHeader
+import com.apollographql.apollo3.api.http.HttpMethod
+import com.apollographql.apollo3.api.http.HttpRequest
+import com.apollographql.apollo3.api.http.HttpResponse
+import com.apollographql.apollo3.exception.ApolloNetworkException
+import io.ktor.client.fetch.AbortController
+import io.ktor.client.fetch.AbortSignal
+import io.ktor.client.fetch.RequestInit
+import io.ktor.client.fetch.fetch
+import io.ktor.http.HttpHeaders
+import io.ktor.util.PlatformUtils
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.await
+import okio.Buffer
+import org.w3c.fetch.Headers
+
+private fun Headers.toHeadersList(): List<HttpHeader> {
+  return js("Array").from(this.asDynamic().entries()).unsafeCast<Array<Array<String>>>().map { HttpHeader(it[0], it[1]) }
+}
+
+private fun <T> buildObject(block: T.() -> Unit): T = (js("{}") as T).apply(block)
+
+private fun AbortController(): AbortController {
+  return if (PlatformUtils.IS_BROWSER) {
+    js("new AbortController()")
+  } else {
+    @Suppress("UNUSED_VARIABLE")
+    val controller = js("eval('require')('abort-controller')")
+    js("new controller()")
+  }
+}
+
+private fun HttpRequest.toRaw(abortSignal: AbortSignal): RequestInit {
+  val bodyBuffer = Buffer()
+
+  val rawHeaders = headers.map { arrayOf(it.name, it.value) }.toMutableList()
+
+  body?.let {
+    it.writeTo(bodyBuffer)
+    rawHeaders.add(arrayOf(HttpHeaders.ContentType, it.contentType))
+  }
+
+  return buildObject {
+    body = bodyBuffer.readUtf8()
+    headers = rawHeaders.toTypedArray()
+    method = when (this@toRaw.method) {
+      HttpMethod.Get -> io.ktor.http.HttpMethod.Get.value
+      HttpMethod.Post -> io.ktor.http.HttpMethod.Post.value
+    }
+    signal = abortSignal
+  }
+}
+
+/**
+ * An [HttpEngine] that calls fetch directly and uses the browser's native JSON parsing
+ * capabilities and then passes the parse JSON into [DynamicJsJsonReader]. In testing, we
+ * see a 4-5x performance improvement using this over [KtorHttpEngine]. On the whole, it is still
+ * ~1/2 the speed of the native JS apollo client.
+ *
+ */
+class FastBrowserJsHttpEngine : HttpEngine {
+  private var disposed = false
+  private val controller = AbortController()
+
+  init {
+    if (!PlatformUtils.IS_BROWSER) {
+      throw RuntimeException("FastHttpEngine is not available outside of the browser.")
```

Comment:
Sounds like this [could work](https://nodejs.org/dist/latest-v18.x/docs/api/globals.html#fetch) on some versions of node too?

Reply:
Yeah, I think it could be fairly easily extended to work generally on node, but figured better leave that until somebody needs it.

Reply:
Makes sense 👍